pkg/cmd/pulumi/state_rename.go

```diff
@@ -128,32 +129,71 @@ func newStateRenameCommand() *cobra.Command {
 	var yes bool
 
 	cmd := &cobra.Command{
-		Use:   "rename <resource URN> <new name>",
+		Use:   "rename [resource URN] [new name]",
 		Short: "Renames a resource from a stack's state",
 		Long: `Renames a resource from a stack's state
 
 This command renames a resource from a stack's state. The resource is specified
-by its Pulumi URN (use ` + "`pulumi stack --show-urns`" + ` to get it) and the new name of the resource.
+by its Pulumi URN and the new name of the resource.
 
 Make sure that URNs are single-quoted to avoid having characters unexpectedly interpreted by the shell.
 `,
 		Example: "pulumi state rename 'urn:pulumi:stage::demo::eks:index:Cluster$pulumi:providers:kubernetes::eks-provider' new-name-here",
-		Args:    cmdutil.ExactArgs(2),
+		Args:    cmdutil.MaximumNArgs(2),
 		Run: cmdutil.RunResultFunc(func(cmd *cobra.Command, args []string) result.Result {
 			ctx := commandContext()
 			yes = yes || skipConfirmations()
-			urn := resource.URN(args[0])
-			newResourceName := args[1]
-			// Show the confirmation prompt if the user didn't pass the --yes parameter to skip it.
-			showPrompt := !yes
 
-			if !urn.IsValid() {
-				return result.Error("The provided input URN is not valid")
+			if len(args) < 2 && !cmdutil.Interactive() {
+				return result.Error("Must supply <resource URN> and <new name> unless in an interactive session")
 			}
 
-			res := runTotalStateEdit(ctx, stack, showPrompt, func(opts display.Options, snap *deploy.Snapshot) error {
-				return stateRenameOperation(urn, newResourceName, opts, snap)
-			})
+			var urn resource.URN
+			var newResourceName tokens.QName
+			switch len(args) {
+			case 0:
+				var snap *deploy.Snapshot
+				err := surveyStack(
+					func() (err error) {
+						urn, err = getURNFromState(ctx, stack, &snap, "Choose a resource to rename:")
+						return
+					},
+					func() (err error) {
+						newResourceName, err = getNewResourceName()
+						return
+					},
+				)
+				if err != nil {
+					return result.FromError(err)
+				}
+
+			case 2: // We got the URN and the name.
+				rName := args[1]
+				if !tokens.IsQName(rName) {
+					reason := "resource names may only contain alphanumerics, underscores, hyphens, dots, and slashes"
+					return result.Errorf("invalid name %q: %s", rName, reason)
+				}
+				newResourceName = tokens.QName(rName)
+			case 1: // We got the urn but not the name
```

Comment:
Don't we need a fallthrough here?

Reply:
Not quite. Each case is actually independent. That said, this was a bug (now fixed).